datafusion/core/src/physical_plan/file_format/delimited_stream.rs

```diff
@@ -0,0 +1,215 @@
+            }
+        });
+
+        let start_offset = match self.remainder.is_empty() {
+            true => 0,
+            false => match record_ends.next() {
+                Some(idx) => {
+                    self.remainder.extend_from_slice(&val[0..idx]);
+                    self.complete
+                        .push_back(Bytes::from(std::mem::take(&mut self.remainder)));
+                    idx
+                }
+                None => {
+                    self.remainder.extend_from_slice(&val);
+                    return;
+                }
+            },
+        };
+        let end_offset = record_ends.last().unwrap_or(start_offset);
+        if start_offset != end_offset {
+            self.complete.push_back(val.slice(start_offset..end_offset));
+        }
+
+        if end_offset != val.len() {
+            self.remainder.extend_from_slice(&val[end_offset..])
+        }
+    }
+
+    /// Marks the end of the stream, delimiting any remaining bytes
+    ///
+    /// Returns `true` if there is no remaining data to be read
+    fn finish(&mut self) -> Result<bool> {
+        if !self.remainder.is_empty() {
+            if self.is_quote {
+                return Err(DataFusionError::Execution(
+                    "encountered unterminated string".to_string(),
+                ));
+            }
+
+            if self.is_escape {
+                return Err(DataFusionError::Execution(
+                    "encountered trailing escape character".to_string(),
+                ));
+            }
+
+            self.complete
+                .push_back(Bytes::from(std::mem::take(&mut self.remainder)))
+        }
+        Ok(self.complete.is_empty())
+    }
+}
+
+impl Iterator for LineDelimiter {
+    type Item = Bytes;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.complete.pop_front()
+    }
+}
+
+/// Given a [`Stream`] of [`Bytes`] returns a [`Stream`] where each
+/// yielded [`Bytes`] contains a whole number of new line delimited records
+pub fn newline_delimited_stream<S>(s: S) -> impl Stream<Item = Result<Bytes>>
+where
+    S: Stream<Item = Result<Bytes>> + Unpin,
+{
+    let delimiter = LineDelimiter::new();
+
+    futures::stream::unfold((s, delimiter), |(mut s, mut delimiter)| async move {
+        loop {
+            if let Some(next) = delimiter.next() {
+                return Some((Ok(next), (s, delimiter)));
+            }
+
+            match s.next().await {
+                Some(Ok(bytes)) => delimiter.push(bytes),
+                Some(Err(e)) => return Some((Err(e), (s, delimiter))),
+                None => match delimiter.finish() {
+                    Ok(true) => return None,
+                    Ok(false) => continue,
+                    Err(e) => return Some((Err(e), (s, delimiter))),
+                },
+            }
+        }
+    })
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use futures::stream::TryStreamExt;
+
+    #[test]
+    fn test_delimiter() {
```

Comment:
Is the case where some `push` doesn't have a delimiter covered? I couldn't find it